stock_available_to_promise_release/models/stock_move.py

```diff
@@ -605,24 +663,77 @@ def unrelease(self, safe_unrelease=False):
             next(iterator)  # skip the current move
             for origin_moves in iterator:
-                origin_moves = origin_moves.filtered(
+                todo_origin_moves = origin_moves.filtered(
                     lambda m: m.state not in ("done", "cancel")
                 )
-                if origin_moves:
-                    origin_moves = move._split_origins(origin_moves)
-                    impacted_picking_ids.update(origin_moves.mapped("picking_id").ids)
+                qty_cancelled_at_step = 0
+                if todo_origin_moves:
+                    qty_to_split = qty_to_unrelease - qty_returned_for_move
+                    moves_to_cancel = move._split_origins(
+                        todo_origin_moves, qty=qty_to_split
+                    )
+                    impacted_picking_ids.update(moves_to_cancel.picking_id.ids)
                     # avoid to propagate cancel to the original move
-                    origin_moves.write({"propagate_cancel": False})
-                    # origin_moves._action_cancel()
-                    moves_to_cancel |= origin_moves
-            moves_to_cancel._action_cancel()
+                    moves_to_cancel.write({"propagate_cancel": False})
+                    moves_to_cancel_for_move |= moves_to_cancel
+                    qty_cancelled_at_step = sum(moves_to_cancel.mapped("product_qty"))
+                # checking that for the current step (pick/pack/ship)
+                # move.product_uom_qty == step.cancelled_qty + move.returned_quanty
+                # If not the case, we have to move back goods in stock.
+                qty_to_return_at_step = (
+                    qty_to_unrelease - qty_cancelled_at_step - qty_returned_for_move
+                )
+                done_moves = origin_moves.filtered(lambda m: m.state == "done")
+                if float_is_zero(qty_to_return_at_step, precision_rounding=rounding):
+                    continue
+                elif not move.rule_id.allow_unrelease_return_done_move:
+                    # Without allow_unrelease_return_done_move enabled,
+                    # only moves that aren't done can be unreleased.
+                    msg_args = {
+                        "move_name": move.name,
+                        "done_move_names": ", ".join(done_moves.mapped("name")),
+                    }
+                    message = _(
+                        (
+                            "You cannot unrelease the move %(move_name)s "
+                            "because some origin moves %(done_move_names)s are done"
+                        ),
+                        **msg_args
+                    )
+                    raise UserError(message)
+                # Multiple pickings can satisfy a move
+                # -> len(move.move_orig_ids.picking_id) > 1
+                # Group done_moves per picking, and create returns
+                groups_iterator = groupby(done_moves, key=lambda m: m.picking_id)
```

Comment:
As shopfloor can create split order, we should create one return per picking type to ensure that all products are in the same return picking. And if the products are in packages, then we face a bug in odoo reservation
```suggestion
                groups_iterator = groupby(done_moves, key=lambda m: m.picking_type_id)
```

Reply:
I tested and it's already better. Still we end-up with one return per move because of the loop on the out moves to unrelease.
The solution would be to not call `moves_to_return._return_quantity_in_stock` here but record each move and quantity to return by picking type. Then at the end, outside the out moves loop, you can call `moves_to_return._return_quantity_in_stock` for all the moves & qty per picking type.